File: trains-gdx-core/src/main/java/com/github/caay2000/trains/render/WorldRender.java
package com.github.caay2000.trains.render;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.github.caay2000.trains.world.City;
import com.github.caay2000.trains.world.Route;
import com.github.caay2000.trains.world.Train;
import com.github.caay2000.trains.world.World;

public class WorldRender {

    private final int screenMargin = 50;

    private final ShapeRenderer shapeRenderer;
    private final SpriteBatch batch;
    private final OrthographicCamera camera;
    float width = 800, height, aspectRatio;


    public WorldRender() {
        shapeRenderer = new ShapeRenderer();
        batch = new SpriteBatch();

        camera = new OrthographicCamera();
        aspectRatio = (float) Gdx.graphics.getHeight() / (float) Gdx.graphics.getWidth();
        height = width * aspectRatio;
        camera.position.set(width / 2, height / 2, 0);
        camera.viewportWidth = width / 2;
        camera.viewportHeight = height / 2;
    }

    public void render(World world) {

        camera.update();

        Gdx.gl.glClearColor(1, 1, 1, 1);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

        Gdx.gl.glEnable(GL20.GL_BLEND);
        Gdx.gl.glBlendFunc(GL20.GL_SRC_ALPHA, GL20.GL_ONE_MINUS_SRC_ALPHA);
        Gdx.gl.glDisable(GL20.GL_BLEND);

        batch.setProjectionMatrix(camera.combined);

        shapeRenderer.setColor(Color.BLACK);
        shapeRenderer.begin(ShapeRenderer.ShapeType.Filled);

        for (City city : world.getCities()) {
            shapeRenderer.circle(
                    city.getPosition().getX() + xOffset(world),
                    city.getPosition().getY() + yOffset(world),
                    city.getPopulation() / 1000 + 1);
        }

        for (Route route : world.getRoutes()) {
            shapeRenderer.line(route.getStart().getPosition().getX() + xOffset(world),
                    route.getStart().getPosition().getY() + yOffset(world),
                    route.getEnd().getPosition().getX() + xOffset(world),
                    route.getEnd().getPosition().getY() + yOffset(world));
        }

        shapeRenderer.setColor(Color.RED);
        for (Train train : world.getTrains()) {
            shapeRenderer.circle(
                    train.getPosition().getX() + xOffset(world),
                    train.getPosition().getY() + yOffset(world),
                    2);
        }

        shapeRenderer.end();
    }

    public int xOffset(World world) {
        return Math.abs(world.getSize().getMinX()) + screenMargin;
    }

    public int yOffset(World world) {
        return Math.abs(world.getSize().getMinY()) + screenMargin;
    }
}

File: trains-gdx-core/src/main/kotlin/com/github/caay2000/trains/render/WorldRender.kt
package com.github.caay2000.trains.render

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.github.caay2000.trains.world.GlobalData
import com.github.caay2000.trains.world.World
import kotlin.math.abs

class WorldRender {

    private val screenMargin = 50
    private val shapeRenderer: ShapeRenderer = ShapeRenderer()
    private val batch: SpriteBatch = SpriteBatch()
    private var width = 800f
    private var height: Float
    private var aspectRatio: Float = Gdx.graphics.height.toFloat() / Gdx.graphics.width.toFloat()
    private val camera: OrthographicCamera = OrthographicCamera(1f, aspectRatio)
    private var worldPopulation = 0
    private var font = BitmapFont()

    fun render(world: World) {
        // debug { "Update delta: ${GlobalData.delta}, elapsed: ${GlobalData.current}" }
        worldPopulation = 0
        camera.update()
        Gdx.gl.glClearColor(1f, 1f, 1f, 1f)
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)
        Gdx.gl.glEnable(GL20.GL_BLEND)
        Gdx.gl.glBlendFunc(GL20.GL_SRC_ALPHA, GL20.GL_ONE_MINUS_SRC_ALPHA)
        Gdx.gl.glDisable(GL20.GL_BLEND)
        batch.projectionMatrix = camera.combined

        shapeRenderer.color = Color.BLACK
        shapeRenderer.begin(ShapeRenderer.ShapeType.Filled)
        for (city in world.cities()) {
            shapeRenderer.circle(
                city.position.x + xOffset(world),
                city.position.y + yOffset(world),
                city.size / 1000 + 1.toFloat()
            )
            worldPopulation += city.size
        }
        for (company in world.companies()) {
            shapeRenderer.color = company.companyColor.color
            for (route in company.routes()) {
                for (i in route.cities.withIndex())
                    if (i.index != 0) {
                        shapeRenderer.line(
                            route.cities[i.index - 1].position.x + xOffset(world),
                            route.cities[i.index - 1].position.y + yOffset(world),
                            route.cities[i.index].position.x + xOffset(world),
                            route.cities[i.index].position.y + yOffset(world)
                        )
                    }
            }
            shapeRenderer.color = Color.RED
            for (train in company.entities()) {
                shapeRenderer.circle(
                    train.position.x + xOffset(world),
                    train.position.y + yOffset(world), 2f
                )
            }
        }

        shapeRenderer.end()

        batch.begin()
        font.draw(batch, "population $worldPopulation", 400f, 400f)
        font.draw(batch, "elapsed ${elapsedString(GlobalData.current)}", 400f, 420f)
        batch.end()
    }

    private fun elapsedString(current: Float): String {
        return "${current.toInt() / 60}:${current.toInt() % 60}"
    }

    private fun xOffset(world: World): Int {
        return abs(world.minX()) + screenMargin
    }

    private fun yOffset(world: World): Int {
        return abs(world.minY()) + screenMargin
    }

    init {
        height = width * aspectRatio
        camera.position[width / 2, height / 2] = 0f
        camera.viewportWidth = width / 2
        camera.viewportHeight = height / 2
    }
}